bosond.cpp

```diff
@@ -99,125 +106,133 @@ int main(int argc, char** argv) {
+            perror("mmap");
+            exit(1);
+        }
+        memset(buffer[i], 0, bufferinfo[i].length);
     }
-    memset(buffer_start, 0, bufferinfo.length);
 
     // Init socket.
     int sock = socket(AF_UNIX, SOCK_STREAM, 0);
 
-    int send_size = bufferinfo.length;
-    if (setsockopt(sock, SOL_SOCKET, SO_SNDBUF, (const void *)&send_size, sizeof(send_size)) < 0) {
-        perror("SETSOCKOPT");
-        exit(1);
-    }
+    if (sendFrames) {
+        int send_size = bufferinfo[0].length;
+        if (setsockopt(sock, SOL_SOCKET, SO_SNDBUF, (const void *)&send_size, sizeof(send_size)) < 0) {
+            perror("SETSOCKOPT");
+            exit(1);
+        }
 
-    struct sockaddr_un addr = { .sun_family = AF_UNIX };
-    strncpy(addr.sun_path, socketPath.c_str(), sizeof(addr.sun_path)-1);
+        struct sockaddr_un addr = { .sun_family = AF_UNIX };
+        strncpy(addr.sun_path, socketPath.c_str(), sizeof(addr.sun_path)-1);
 
-    if (connect(sock, (sockaddr*) (&addr), sizeof(addr)) < 0) {
-        perror("CONNECT");
-        exit(1);
-    }
+        if (connect(sock, (sockaddr*) (&addr), sizeof(addr)) < 0) {
+            perror("CONNECT");
+            exit(1);
+        }
 
-    std::ostringstream headers;
-    headers << "Brand: flir\n";
-    headers << "Model: boson\n";
-    headers << "ResX: " << width << '\n';
-    headers << "ResY: " << height << '\n';
-    headers << "FPS: 60\n";
-    headers << "FrameSize: " << (width * height * 2) << '\n';
-    headers << "PixelBits: 16\n";
-    headers << '\n';
-    auto header_str = headers.str();
-    if (sendall(sock, header_str.data(), header_str.length()) < 0) {
-        perror("HEADERS");
-        exit(1);
+        std::ostringstream headers;
+        headers << "Brand: flir\n";
+        headers << "Model: boson\n";
+        headers << "ResX: " << width << '\n';
+        headers << "ResY: " << height << '\n';
+        headers << "FPS: 60\n";
+        headers << "FrameSize: " << (width * height * pix_bytes) << '\n';
+        headers << "PixelBits: " << (pix_bytes * 2) << '\n';
+        headers << '\n';
+        auto header_str = headers.str();
+        if (sendall(sock, header_str.data(), header_str.length()) < 0) {
+            perror("HEADERS");
+            exit(1);
+        }
     }
 
     // Activate streaming
-    int type = bufferinfo.type;
-    if(ioctl(fd, VIDIOC_STREAMON, &type) < 0){
+    if (ioctl(fd, VIDIOC_STREAMON, &bufferinfo[0].type) < 0) {
         perror("VIDIOC_STREAMON");
         exit(1);
     }
 
-    system_clock::time_point t0 = system_clock::now();
+
+    steady_clock::time_point t0 = steady_clock::now();
+
+    // Indices of the buffer being filled (i_active) and the buffer
+    // that is full and is ready to send (i_full)
+    int i_active = 0;
+    int i_full = 1;
+
+    // Put the first buffer in the incoming queue.
+    if (ioctl(fd, VIDIOC_QBUF, &bufferinfo[i_active]) < 0) {
+        perror("VIDIOC_QBUF");
+        exit(1);
+    }
+
```

Comment:
Does it not make sense to queue all buffers first?

Reply:
Maybe ... I just wanted one buffer queued while the other is sending. I'll add an issue about queuing more than one buffer to see if that improves anything.